### PR TITLE
[Speedy IPP] Cache IPP onboarding result

### LINKED ISSUE #8642: [Mobile Payments] Cache onboarding result
* We cache only Onboarding Completed State
* Run onboarding check in the background on every app start. Make sure to run it only for the users who successfully used IPP before in the last 30 days. Also, make sure to run the check-in atomic way, so if a user after the app's restart quickly starts IPP flow we won't mess up the cache with the results for 2 checks that run in parallel. Fetch only for the selected site!
* We can use it in the memory cache
* We invalidate cache:
  * If any error comes from the SDK or the API during the payment flow
  * Plugin Selection clicked
  * Site change
  * Logout

pdfdoF-2n0-p2

### PR DESCRIPTION
**Pull request by kidinov**
Closes: #8642

### Description

The PR adds in-memory caching on the IPP onboarding check result

* We cache only Onboarding Completed State
* Run onboarding check in the background on every app start. Make sure to run it only for the users who successfully used IPP before in the last 30 days. Also, make sure to run the check-in atomic way, so if a user after the app’s restart quickly starts IPP flow we won’t mess up the cache with the results for 2 checks that run in parallel. Fetch only for the selected site!
* We can use it in the memory cache
* We invalidate cache:
  * If any error comes from the SDK or the API during the payment flow
  * Plugin Selection clicked
  * Site change
  * Logout

@samiuelson @AnirudhBhat this is important change so I would like to ask both of you to review this

### Testing instructions

* Open the app and make a payment. Notice that the next payment will go faster because we use the cached value
* If payment (any IPP or TTP) payment was done within the last 30 days, notice that there is no loading on the Payment Hub screen
* Change data on the phone on 31 days and notice that when Hub is open then we do an onboarding check 
* Try different combinations of the cases when the cache is not valid anymore. E.g. open the app with onboarding completed, then go to the web and make changes to make onboarding not completed. Try to make payment. Notice that cache is invalidated in this case and next payment we will fetch the onboarding status
* Do the same, but with logout
* Do the same, but with a site change
* Do the same, but with "plugin selection" click on the hub screen

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/extensions/NumberExt.kt [hunk 1]
```diff
@@ -33,4 +33,8 @@ infix fun <T> Comparable<T>?.greaterThan(other: T) =
     this?.let { it > other }
         ?: false
 
+infix fun <T> Comparable<T>?.lesserThan(other: T) =
```
Comment (samiuelson) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/extensions/NumberExt.kt:36]:
```suggestion
infix fun <T> Comparable<T>?.smallerThan(other: T) =
```
I'd opt for renaming it to `smallerThan`.

**Review by samiuelson — APPROVED** [review]
I tested the branch and it works as expected. Just left a minor comment. Also, I'd prefer @AnirudhBhat to have another look at it as well.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/hub/CardReaderHubViewModel.kt [hunk 2]
```diff
@@ -317,6 +316,7 @@ class CardReaderHubViewModel @Inject constructor(
     }
 
     private fun onCardReaderPaymentProviderClicked() {
+        cardReaderChecker.invalidateCache()
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/hub/CardReaderHubViewModel.kt:352]: ❓ There seems to be a test missing for this. What do you think about adding one?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/onboarding/CardReaderOnboardingChecker.kt [hunk 3]
```diff
@@ -76,24 +74,33 @@ class CardReaderOnboardingChecker @Inject constructor(
...
+            cachedValue.state
+        } else fetchOnboardingState(pluginType)
+            .also { state ->
+                val (status, version) = when (state) {
+                    is OnboardingCompleted -> {
+                        cardReaderOnboardingCheckResultCache.value =
+                            CardReaderOnboardingCheckResultCache.Result.Cached(state)
+                        CARD_READER_ONBOARDING_COMPLETED to state.version
+                    }
+                    is CashOnDeliveryDisabled -> CARD_READER_ONBOARDING_PENDING to state.version
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/onboarding/CardReaderOnboardingChecker.kt:94]: ❓ What do you think about adding tests to verify that the cache value is `NotCached` for all the states of onboarding except for `ONBOARDING_COMPLETED`? This way, we catch if there is any unintentional change in the logic of the caching.

**Review by AnirudhBhat — APPROVED** [review]
Great work on this @kidinov. I've left a couple of minor comments that are non-blocking for the merge.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/extensions/NumberExt.kt (hunk 1 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/extensions/NumberExt.kt:36]: Yeah, I am not an English expert but I ran "what is opposite to greater" in google and chatgpt and the result was that, so I guess this is grammatically correct =)

image
image

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/hub/CardReaderHubViewModel.kt (hunk 2 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/hub/CardReaderHubViewModel.kt:352]: 👍

61d37501307f7494a4bc41b6cd18be8f1cca5f6b

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/onboarding/CardReaderOnboardingChecker.kt (hunk 3 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/onboarding/CardReaderOnboardingChecker.kt:94]: Done 1d56849de09b46e4616435509530965e85f231ef

**Pull request merged**